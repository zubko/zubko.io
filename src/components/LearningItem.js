/**
 * Item component for Learning page
 *
 * @flow
 */

import React, { useState } from 'react';

import styled from '@emotion/styled';

import Link from './Link';
import { rhythm } from '../Typography';
import { Styles } from '../Theme';

type Props = {
  node: {
    html: string,
    frontmatter: { author: string, title: String, year: string },
  },
};

const LearningItem = ({ node }: Props) => {
  const {
    html,
    frontmatter: { author, title, year },
  } = node;
  const [opened, setOpened] = useState(false);
  const handleClick = event => {
    event.preventDefault();
    setOpened(!opened);
  };
  return (
    <Container>
      <Triangle opened={opened} onClick={handleClick} />
      <Main>
        <StyledLink to="#" usePlainStyle onClick={handleClick}>
          {author} - {title}
          {year ? ` '${year}` : ''}
        </StyledLink>
        {opened && (
          <p>
            <div
              css={Styles.markdownInline}
              dangerouslySetInnerHTML={{ __html: html }}
            />
          </p>
        )}
      </Main>
    </Container>
  );
};

export default LearningItem;

const Container = styled.li`
  display: flex;
  flex-direction: row;
`;
const Main = styled.div`
  flex: 1;
`;

const StyledLink = styled(Link)`
  background-image: none;
`;

const Triangle = styled.div`
  display: inline-block;
  margin-right: ${rhythm(0.5)};
  transform: rotate(${props => (props.opened ? 90 : 0)}deg);
  transform-origin: 50% 45%;
  align-self: flex-start;
  transition: transform 0.2s;
  cursor: pointer;
  user-select: none;
`;
Triangle.defaultProps = {
  children: 'â–¶',
};
