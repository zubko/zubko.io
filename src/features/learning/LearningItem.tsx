/**
 * Item component for Learning page
 */

import { MouseEvent, useState } from "react";

import styled from "styled-components";

import { Link } from "../../components/Link";
import { Markdown } from "../../components/Markdown";
import { rhythm } from "../../Typography";

type Props = {
  node: {
    html: string;
    frontmatter: { author: string; title: string; year: string };
  };
};

export const LearningItem = ({ node }: Props) => {
  const {
    html,
    frontmatter: { author, title, year },
  } = node;
  const [opened, setOpened] = useState(false);
  const handleClick = (
    event: MouseEvent<HTMLDivElement | HTMLAnchorElement>
  ) => {
    event.preventDefault();
    setOpened(!opened);
  };
  return (
    <Container>
      <Triangle opened={opened} onClick={handleClick} />
      <TitleAndContent>
        <StyledLink to="#" usePlainStyle onClick={handleClick}>
          {author} - {title}
          {year ? ` '${year}` : ""}
        </StyledLink>
        {opened && (
          <Content>
            <Markdown htmlContent={html} inline />
          </Content>
        )}
      </TitleAndContent>
    </Container>
  );
};

const Container = styled.li`
  display: flex;
  flex-direction: row;
`;

const Triangle = styled.div.attrs({
  children: "â–¶",
})<{ opened: boolean }>`
  display: inline-block;
  margin-right: ${rhythm(0.5)};
  transform: rotate(${(props) => (props.opened ? 90 : 0)}deg);
  transform-origin: 50% 45%;
  align-self: flex-start;
  transition: transform 0.2s;
  cursor: pointer;
  user-select: none;
`;

const TitleAndContent = styled.div`
  flex: 1;
`;

const StyledLink = styled(Link)`
  background-image: none;
`;

const Content = styled.div`
  && {
    /* '&&' to override some list related styles in Typography.js
       which made every last sub element of 'li' to not have margin-bottom */
    margin: ${rhythm(0.5)} 0;
  }
`;
