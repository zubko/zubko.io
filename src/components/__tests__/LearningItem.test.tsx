import React from "react";
import renderer from "react-test-renderer";

import LearningItem from "../LearningItem";

describe("LearningItem", () => {
  it("renders without regressions", () => {
    const tree = renderer
      .create(
        <LearningItem
          node={{
            html: "<p>some html</p>",
            frontmatter: { author: "author", title: "title", year: 2019 },
          }}
        ></LearningItem>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
