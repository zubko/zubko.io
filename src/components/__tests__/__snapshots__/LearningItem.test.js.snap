// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LearningItem renders without regressions 1`] = `
<li
  className="css-1mcfn8a-Container exa79cs0"
>
  <div
    className="css-h2lzyt-Triangle exa79cs3"
    onClick={[Function]}
  >
    â–¶
  </div>
  <div
    className="css-vx98lc-Main exa79cs1"
  >
    <a
      className="exa79cs2 css-1i4fh7c-linkPlain-StyledLink"
      href="#"
      onClick={[Function]}
    >
      author
       - 
      title
       '2019
    </a>
  </div>
</li>
`;
